### PR TITLE
Implement Continual-MEGA baseline and benchmark

### PR DESCRIPTION
**Pull request by krzkapera**

### CODE REVIEW

File: src/pyclad/vision/models/continual_mega_baseline/continual_mega_baseline.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        self,
+        data: np.ndarray,
+        labels: np.ndarray,
+        masks: Optional[np.ndarray] = None,
+        on_epoch_end: Optional[Callable[[int, float], None]] = None,
+    ) -> None:
+        images = self._to_tensor(data)
+        targets = torch.from_numpy(labels)
+        pixel_targets = self._mask_tensor(masks, labels)
+        while self.trainer.epochs_completed < self.config.epochs:
```
Comment (Nyderx) [inline src/pyclad/vision/models/continual_mega_baseline/continual_mega_baseline.py:80]: **Bug: `fit()` trains only on the first concept, then becomes a no-op.**

`ContinualMegaBaselineTrainer.epochs_completed` is set once in `__init__` (trainer.py:35) and only ever incremented in `train_epoch` (trainer.py:53) — nothing resets it between calls to `fit()`.

`while self.trainer.epochs_completed < self.config.epochs` runs `config.epochs` iterations on the first training concept and ends with `epochs_completed == config.epochs`. On every subsequent call to `fit()` (one per remaining concept in the scenario, e.g. ~20 for Continual-MEGA scenario 2), the loop condition is false immediately, so `train_epoch()` never runs again. The model silently stops learning after the first concept, defeating the continual-learning benchmark.

Looks like `epochs_completed` needs to be reset to 0 at the start of each `fit()` call (or renamed/tracked per-call rather than cumulatively).

File: src/pyclad/vision/models/continual_mega_baseline/continual_mega_baseline.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        )
+
+    def _mask_tensor(self, masks: Optional[np.ndarray], labels: np.ndarray) -> torch.Tensor:
+        if masks is None:
+            if labels.any():
+                raise ValueError(f"{self.name()} needs ground truth masks to train on anomalous samples")
+            return torch.zeros((len(labels), self.config.image_size, self.config.image_size))
+        return torch.from_numpy(masks).float()
+
+    def _to_tensor(self, images: np.ndarray) -> torch.Tensor:
```
Comment (Nyderx) [inline src/pyclad/vision/models/continual_mega_baseline/continual_mega_baseline.py:118]: **Bug: no CLIP mean/std normalization applied to model inputs.**

`_to_tensor()` only scales pixels to [0,1] via `div_(255.0)`; the standard per-channel CLIP normalization (mean/std) is never applied anywhere in this pipeline (checked `backbone.py`, `visual_encoder.py`, and this file).

Since the backbone is frozen (`requires_grad_(False)`, loaded via `open_clip.load_openai_model`), it receives out-of-distribution inputs on every forward pass in both `fit()` and `predict()`, which will degrade visual features and downstream anomaly scores for the whole model.

Suggest applying the OpenAI CLIP normalization constants (mean ≈ [0.481, 0.458, 0.408], std ≈ [0.269, 0.261, 0.276]) after the [0,1] scaling in `_to_tensor`.

File: src/pyclad/vision/data/lazy_concepts.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    def __getitem__(self, index: int) -> Concept:
+        name, samples = self._samples_by_concept[index]
+        return build_vision_concept(
+            name=name,
+            samples=samples,
+            options=self._options,
+            with_labels=self._with_labels,
+        )
+
+    def __iter__(self) -> Iterator[Concept]:
```
Comment (Nyderx) [inline src/pyclad/vision/data/lazy_concepts.py:36]: **Efficiency: test set is re-decoded from disk once per training concept.**

`LazyVisionConceptList.__iter__` re-runs `build_vision_concept` → `materialize_samples` (disk read + resize) for every sample on every iteration. `SupervisedConceptIncrementalScenario.run()` iterates `self._dataset.test_concepts()` inside the per-training-concept outer loop, so for a ~20-concept scenario the full test set gets reloaded from disk ~20x more than necessary compared to eagerly materializing it once.

Might be worth caching materialized concepts (or the decoded arrays) across outer-loop iterations if this is a bottleneck in practice.

File: src/pyclad/vision/data/base.py [hunk 4]
```diff
@@ -51,10 +74,11 @@ def read_dataset(
         data_mode: str = "numpy",
         resize_to: Optional[Tuple[int, int]] = None,
         color_mode: str = "rgb",
+        interpolation: str = "bilinear",
+        apply_exif_transpose: bool = False,
         max_train_samples_per_category: Optional[int] = None,
         max_test_samples_per_category: Optional[int] = None,
     ) -> ConceptsDataset:
-        validate_read_options(data_mode=data_mode, color_mode=color_mode)
         samples = self.index_samples(
```
Comment (Nyderx) [inline src/pyclad/vision/data/base.py:82]: **Nit: invalid read options now fail after indexing instead of before.**

`read_dataset()` calls `self.index_samples(...)` here before `ImageLoadOptions` is constructed (its `__post_init__` is what validates `data_mode`/`color_mode`/`interpolation`). An invalid option value now only surfaces after the full (potentially expensive) directory scan/indexing pass completes, rather than failing fast.

Consider validating `ImageLoadOptions` before calling `index_samples`.

File: src/pyclad/vision/data/base.py [hunk 5]
```diff
@@ -169,23 +180,19 @@ def select_categories(
     return list(requested_categories)
 
 
-def list_image_files(directory: Path, image_extensions: Iterable[str]) -> List[Path]:
+def list_image_files(directory: Path, image_extensions: Iterable[str], recursive: bool = False) -> List[Path]:
```
Comment (Nyderx) [inline src/pyclad/vision/data/base.py:183]: **Nit: unused `recursive` parameter.**

`list_image_files` gained a `recursive` param, but no caller in the codebase (checked all 9 call sites in `readers.py` and `generic_reader.py`) ever passes `recursive=True`, so the `directory.rglob('*')` branch is dead/untested code. Either wire it up where it's needed or drop it until there's a caller.

File: src/pyclad/vision/models/continual_mega_baseline/continual_mega_baseline.py (hunk 2 above)
Comment (krzkapera) [reply src/pyclad/vision/models/continual_mega_baseline/continual_mega_baseline.py]: Brak normalizacji CLIP — NIETRAFIONE, to celowa wierność

To nie jest przeoczenie, tylko odtworzenie referencji. Dowody:

- dataset/continual.py w ścieżce ewaluacji robi wyłącznie `convert("RGB")`, `exif_transpose`, `Resize(336, BICUBIC)` i `ToTensor()` — bez `Normalize`.
- CLIP/clip.py przypisuje `model.visual.image_mean/image_std`, ale `grep` po całym repozytorium referencji pokazuje, że nikt tych pól nigdy nie czyta.
- Rozstrzygające empirycznie: z wydanych przez autorów checkpointów nasz ewaluator odtwarza liczby z paperu dokładnie (Image ACC 76.8, Pixel ACC 27.5). Gdybyśmy podawali im znormalizowane wejście, dostalibyśmy inne liczby — te checkpointy były wytrenowane na zakresie [0,1].

File: src/pyclad/vision/data/lazy_concepts.py (hunk 3 above)
Comment (krzkapera) [reply src/pyclad/vision/data/lazy_concepts.py]: Test set dekodowany raz na koncept treningowy — TRAFNA OBSERWACJA, ale nie do naprawy tak, jak sugeruje

Obserwacja jest prawdziwa. Propozycja „cache materialized concepts" jest jednak niewykonalna dla tego benchmarku: scenariusz 2 ma 95 tys. obrazów testowych, co przy 336×336×3 uint8 daje 32 GB w pamięci. Leniwe wczytywanie istnieje właśnie po to.

Dlatego nasze ewaluacje nie chodzą przez scenariusz — eval_adct.py liczy koncept po koncepcie i zapisuje wyniki przyrostowo. Zostawiam bez zmian; to świadomy kompromis, nie niedopatrzenie.

File: src/pyclad/vision/data/base.py (hunk 4 above)
Comment (krzkapera) [reply src/pyclad/vision/data/base.py]: powinno być poprawione, możesz jeszcze sprawdzić, co twój na rozwiązanie powie
